companion.py

```diff
@@ -355,33 +391,39 @@ def query(self, endpoint):
             self.retrying = False
             self.login()
             raise CredentialsError()
+
         elif 500 <= r.status_code < 600:
             # Server error. Typically 500 "Internal Server Error" if server is down
             self.dump(r)
-            raise ServerError()
+            raise ServerError('Received error {} from server'.format(r.status_code))
 
         try:
-            r.raise_for_status()	# Typically 403 "Forbidden" on token expiry
-            data = r.json()		# May also fail here if token expired since response is empty
-        except:
+            r.raise_for_status()  # Typically 403 "Forbidden" on token expiry
+            data = r.json()  # May also fail here if token expired since response is empty
+
+        except (requests.HTTPError, ValueError) as e:
             print_exc()
             self.dump(r)
             self.close()
+
             if self.retrying:		# Refresh just succeeded but this query failed! Force full re-authentication
                 self.invalidate()
                 self.retrying = False
                 self.login()
-                raise CredentialsError()
+                raise CredentialsError('query failed after refresh') from e
+
             elif self.login():		# Maybe our token expired. Re-authorize in any case
                 self.retrying = True
                 return self.query(endpoint)
+
             else:
                 self.retrying = False
-                raise CredentialsError()
+                raise CredentialsError('Invalid JSON or HTTP error') from e
```

Comment:
This can be read as `Invalid (JSON or HTTP error)`.  Swap them around -> "HTTP error or invalid JSON".